README
Net-Twitter-FriendFinder

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc Net::Twitter::FriendFinder



Changes
Revision history for Net-Twitter-FriendFinder

0.03       2007-11-16/22:18
           - add mask() method.
          
0.02       2007-11-15/00:18
           - first release to CPAN

0.01_08    2007-11-14/19:29
           - use Net::Twitter::Diff insetead of Net::Twitter

0.01_07    2007-11-14/00:19
           - Update Net::Twitter::FriendFinder::From::TwitterDiff for version up.

0.01_06    2007-11-09/12:50
           - Add Net::Twitter::FriendFinder::From::URL.

0.01_05    2007-01-09/11:33
           - Add Net::Twitter::FriendFinder::From::TwitterDiff.

0.01_04    2007-11-08/16:40
           - Add Net::Twitter::FriendFinder::From::TwitterKensaku.

0.01_03    2007-11-08/15:04
           - filter_already_followers support. it can be reduced API request.
             for this fix you must set username and passowrd at new() parameters
           - support sleep setting. 

0.01_02    2007-11-08/00:21
           - Bug Fix: can not load custom resource module.

0.01_01    2007-11-07/23:31
           -First version, released on an unsuspecting world.



lib/Net/Twitter/FriendFinder/From/URL.pm
package Net::Twitter::FriendFinder::From::URL;

use strict;
use warnings;
use base qw/Net::Twitter::FriendFinder::From/;
use Web::Scraper;
use URI;

our $ORDER_SCORE = 30;

sub search {
    my $self = shift;
    my $url  = shift;

    my $handicap = $self->{handicap};

    my $scrap = scraper {
        process "a",
        'url[]' => '@href',
    };

    my $res = $scrap->scrape( URI->new( $url ) );

    my $data = {};
    my $order_point = $ORDER_SCORE;
    for my $url ( @{ $res->{url} } ) {
        if( $url->as_string =~ /^(http|https):\/\/(www\.|)twitter\.com/  ) {
             my ( $name ) = $url->as_string =~ m|twitter.com/(\w+)|;
            # order point + prev point
            my $score = $order_point ;
            $score +=  $data->{ $name } ? $data->{ $name }+1 : 1 ;
            $score = $score * $handicap ;
            $data->{ $name } = $score;
            $order_point-- if $order_point > 0 ;
        }
    }

    return $data;
}

1;

=head1 NAME

Net::Twitter::FriendFinder::From::URL - Find Twitter Friend from URL

=head1 SYNOPSYS

use Net::Twitter::FriendFinder;
 my $tf 
    = Net::Twitter::FriendFinder->new({ 
        from => {
            'URL' => {},
        }
      });
    $tf->search( 'http://twitter.g.hatena.ne.jp/keyword/Friends%e7%99%bb%e9%8c%b2%e3%81%94%e8%87%aa%e7%94%b1%e3%81%ab%20Part3' );
    $tf->show();

=head1 DESCRIPTION

seach your choice link and try to find a tag which has link to twitter pgae.

=head1 MODULE

=head2 search

set url you want to search.

=head1 AUTHOR

Tomohiro Teranishi<tomohiro.teranishi@gmail.com>

=cut


lib/Net/Twitter/FriendFinder/From/Google.pm
package Net::Twitter::FriendFinder::From::Google;

use strict;
use warnings;
use base qw/Net::Twitter::FriendFinder::From/;
use Web::Scraper;
use URI;
use URI::Escape;
__PACKAGE__->mk_accessors(qw/lang/);

my $url = "http://www.google.com/search?q=%s+site:.twitter.com&num=%d&start=%d&hl=%s";
our $ORDER_SCORE = 30;

sub search {
    my $self     = shift;
    my $keyword  = shift;

    my $coverage = $self->{coverage};
    my $handicap = $self->{handicap};
    my $lang     = $self->{lang} || '';

    $keyword = URI::Escape::uri_escape( $keyword );
    $keyword =~ s/%20/+/g;
    my $google_url = scraper {
        process "table> tr > td.j > font > span.a",
        description => 'TEXT',
    };

    my $google = scraper {
        process "div.g",
        "urls[]" => $google_url;
    };

    my $data = {};
    my $order_point = $ORDER_SCORE;
    for( my $page = 1 ; $page <= $coverage ; $page++ ) {
        my $target_uri = sprintf( $url , $keyword , 20 , ( $page-1 ) * 20  , $lang )  ;
        my $res = $google->scrape( URI->new( $target_uri ) );

        for my $item ( @{ $res->{urls} } ) {
            my ( $name ) = $item->{description} =~ m|twitter.com/(\w+)/|;
            next unless defined $name;
            # order point + prev point
            my $score = $order_point ;
            $score +=  $data->{ $name } ? $data->{ $name }+1 : 1 ;
            $score = $score * $handicap ;
            $data->{ $name } = $score;
            $order_point-- if $order_point > 0 ;
        }
    }
    return $data;
}

1;

=head1 NAME

Net::Twitter::FriendFinder::From::Google - Search from google.

=head1 SYNOPSYS

   use strict;
   use warnings;
   use FindBin::libs;
   use Net::Twitter::FriendFinder;
   
   
    my $tf 
       = Net::Twitter::FriendFinder->new({ 
            from => {
               Google  => { coverage => 5 , lang => 'ja' },
            }
         });
   
       $tf->search('perl');
       $tf->show();

=head1 DESCRIPTION

Find twitter friends from google search.  

Internally run google search like below.
  
  site:.twitter.com [your keyword]

=head1 coverage

 coverage = 20 entry.

So if you set 

 coverage => 5,

then 20 * 5 = 100. means try to find twitter friend from 100 entry.

=head1 MODULE

=head2 search

=head1 AUTHOR

Tomohiro Teranishi<tomohiro.teranishi@gmail.com>

=cut


lib/Net/Twitter/FriendFinder/From/Twitter.pm
package Net::Twitter::FriendFinder::From::Twitter;

use strict;
use warnings;
use base qw/Net::Twitter::FriendFinder::From/;
use Web::Scraper;
use URI;
use URI::Escape;

our $ORDER_SCORE = 30;
my $url = "http://twitter.com/tw/search/users?q=%s&page=%d";

sub search {
    my $self     = shift;
    my $keyword  = shift;

    my $coverage = $self->{coverage};
    my $handicap = $self->{handicap};

    $keyword = URI::Escape::uri_escape( $keyword );
    $keyword =~ s/%20/+/g;

    my $ws0 = scraper {
        process "div.user_search > div > a.screen_name > span.detail",
        description => 'TEXT',
    };
    my $ws1 = scraper {
        process "div#results > div.user_search",
        'ids[]' => $ws0,
    };

    my $data = {}; 
    my $order_point = $ORDER_SCORE;
    for( my $page = 1 ; $page <= $coverage ; $page++ ) {
        my $target_uri = sprintf( $url , $keyword , $page )  ;
        my $res = $ws1->scrape( URI->new( $target_uri ) );

        for my $item ( @{ $res->{ids} } ) {
            my  $name  = $item->{description};
            next unless defined $name;
            my $score = $order_point * $handicap ;
            $data->{ $name } = $score;
            $order_point-- if $order_point > 0 ;
            
        }
    }

    return $data;
}

1;

=head1 NAME

Net::Twitter::FriendFinder::From::Twitter - Search from twitter people search.

=head1 SYNOPSYS


   use strict;
   use warnings;
   use FindBin::libs;
   use Net::Twitter::FriendFinder;
   
   
    my $tf 
       = Net::Twitter::FriendFinder->new({ 
            from => {
               Twitter => { coverage => 5 },
            }
         });
   
       $tf->search('perl');
       $tf->show();

=head1 DESCRIPTION

Find twitter friends from twitter people search 

e.g.  http://twitter.com/tw/search/users?q=perl

=head1 coverage

 coverage = 20 entry.

So if you set 

 coverage => 5,

then 20 * 5 = 100. means try to find twitter friend from 100 entry.

=head1 MODULE

=head2 search

=head1 AUTHOR

Tomohiro Teranishi<tomohiro.teranishi@gmail.com>

=cut


lib/Net/Twitter/FriendFinder/From/TwitterDiff.pm
package Net::Twitter::FriendFinder::From::TwitterDiff;

use strict;
use warnings;
use Net::Twitter::Diff;
use base qw/Net::Twitter::FriendFinder::From/;
__PACKAGE__->mk_accessors(qw/username password/);

our $SCORE = 50;

sub search {
    my $self    = shift;
    my $keyword = shift;
    my $diff = Net::Twitter::Diff->new( username => $self->{username} , password => $self->{password} );

    my $screen_names = [];
    if( !defined $keyword ) { 
        my $res = $diff->diff();
        $screen_names = $res->{not_following};
    }
    else {
        my $res = $diff->comp_following( $keyword );
        $screen_names = $res->{not_me};
    }

    my $data = {};
    for my $screen_name ( @{ $screen_names } ) {
        $data->{ $screen_name } = $SCORE * $self->{handicap};
    }

    return $data;

}

1;

=head1 NAME

Net::Twitter::FriendFinder::From::TwitterDiff - search from Net::Twitter::Diff

=head1 SYNOPSYS

    use Net::Twitter::FriendFinder;
    my $tf 
    = Net::Twitter::FriendFinder->new({ 
        from => {
            'TwitterDiff' => { username => 'tomyhero' , password => '********' }
        }
      });

    # this one search who you are not following but followed
    $tf->search();
    or 
    # If you set parameter then , search $twitter_name's following but you are not following
    # $tf->search( $twitter_name );
    $tf->show();

=head1 DESCRIPTION

You can get Net::Twitter::Diff from bellow svn repo for now.

 svn co http://svn.coderepos.org/share/lang/perl/Net-Twitter-Diff/trunk/

Search friends who are not folloing bug they are following you.
Search foloowing somebody has.

=head1 METHOD

=head2 search

=head1 AUTHOR

Tomohiro Teranishi<tomohiro.teranishi@gmail.com>

=cut


lib/Net/Twitter/FriendFinder/From/TwitterKensaku.pm
package Net::Twitter::FriendFinder::From::TwitterKensaku;

use strict;
use warnings;
use base qw/Net::Twitter::FriendFinder::From/;
use Web::Scraper;
use URI;
use URI::Escape;

our $ORDER_SCORE = 30;
my $url = 'http://twitter.1x1.jp/search/?keyword=%s&offset=%d';

sub search {
    my $self     = shift;
    my $keyword  = shift;

    my $coverage = $self->{coverage};
    my $handicap = $self->{handicap};

    $keyword = URI::Escape::uri_escape( $keyword );
    $keyword =~ s/%20/+/g;

    my $ws0 = scraper {
        process "div.pagination > table.list > tbody > tr",
            'list[]' => scraper {
                process '//td[2]', id => 'TEXT';
             },
    };

    my $data = {}; 
    my $order_point = $ORDER_SCORE;
    my $prev_top = '';
    for( my $page = 1 ; $page <= $coverage ; $page++ ) {
        my $target_uri = sprintf( $url , $keyword , $page )  ;
        my $res = $ws0->scrape( URI->new( $target_uri ) );
        
        for my $item ( @{ $res->{list} } ) {
            my  ( $name )  = $item->{id} =~ m|@(\w+)|;
            next unless defined $name;

            # order point + prev point
            my $score = $order_point ;
            $score +=  $data->{ $name } ? $data->{ $name }+1 : 1 ;
            $score = $score * $handicap ;
            $data->{ $name } = $score;
            $order_point-- if $order_point > 0 ;
            
        }
    }

    return $data;
}


1;

=head1 NAME

Net::Twitter::FriendFinder::From::TwitterKensaku - TwitterKensaku

=head1 SYNOPSYS

 use Net::Twitter::FriendFinder;
 my $tf 
    = Net::Twitter::FriendFinder->new({ 
        from => {
            'TwitterKensaku' => { coverage => 2 },
        }
      });

    $tf->search( 'perl' );
    $tf->show();

=head1 DESCRIPTION

get data from http://twitter.1x1.jp/search/ 

This is nice site.

=head1 METHOD

=head2 search

=head1 AUTHOR

Tomohiro Teranishi<tomohiro.teranishi@gmail.com>

=cut


lib/Net/Twitter/FriendFinder.pm
package Net::Twitter::FriendFinder;

use warnings;
use strict;
use UNIVERSAL::require;
use base qw/Class::Accessor::Fast/;
__PACKAGE__->mk_accessors(qw/default from setting ids scores resources filters mask/);
use Text::SimpleTable;
use Net::Twitter::Diff;

our $VERSION = '0.03';


sub search {
    my $self = shift;
    my $keyword = shift;

    $self->{keyword} = $keyword;
    # set default of default :-p
    my $default = {};
    $default->{coverage}    = $self->{default}{coverage} || 1;

     $self->{filters} = [];

    my @resources = keys %{ $self->{from} };
    my $data = {}; 
    $self->{resources} = ();
    for my $resource ( @resources ) {
        my $module = substr( $resource , 0 , 1 ) eq '+' ? substr( $resource ,1): "Net::Twitter::FriendFinder::From::" . $resource;
        push @{ $self->{resources} } , $module ;
        $module->require or die $@;
        my $conf =  $self->{from}{ $resource } ;
        $conf->{coverage} = $conf->{coverage} || $default->{coverage};
        $conf->{handicap} = $conf->{handicap} || 1;

        my $from = $module->new( $conf );
        my $results =  $from->search( $keyword );

        foreach my $id ( keys %{ $results } ) {
            $data->{ $id } = $results->{ $id } + ( $data->{ $id } || 0 );
        }
    }

    if( $self->{setting}{filter_already_followers} ) {
        my $twitter = Net::Twitter::Diff->new( username=> $self->{setting}{username} , password => $self->{setting}{password} );
        my $xfollowing = $twitter->xfollowing;
    
        my @filters = ();

        for my $item ( @{ $xfollowing } ) {
            my $screen_name = $item->{screen_name} ;
            if( defined $data->{ $screen_name } ) {
                push @filters , { id => $screen_name , score =>  $data->{ $screen_name } };
                delete $data->{ $screen_name };
            }
        }
        $self->{filters} = \@filters;

    }
    if( defined $self->{setting}{username} ) {
        delete $data->{  $self->{setting}{username} } if defined  $data->{  $self->{setting}{username} };
    }
    for my $screen_name ( @{ $self->{mask} } ) {
        delete $data->{ $screen_name } if defined $data->{ $screen_name };
    }

    my @ids = sort { $data->{ $a } <=> $data->{ $b } } keys %{ $data };

    my $limit = $self->{setting}{limit} || scalar @ids;
    @ids = reverse @ids;
    if( scalar @ids > $limit ) {
        @ids = @ids[0...$limit-1];
    }
    $self->{ids} = \@ids;
    $self->{scores} = $data;

    return 1;
}

sub show {
    my $self = shift;

    my $t = Text::SimpleTable->new( 44 );
    $t->row( 'Resources' );
    print $t->draw;

    my $t0 = Text::SimpleTable->new( 44 );
    for my $resource ( @{ $self->{resources} } ) {
        $t0->row( $resource );
    }
    print $t0->draw;

    my $t1 = Text::SimpleTable->new( 7 , 34 );
    $t1->row('Keyword' , $self->{keyword} );
    print $t1->draw;

    my $t2 = Text::SimpleTable->new( 3, 20, 15 );
    $t2->row( '#', 'Twitter id' , 'Score' );
    print $t2->draw;
    
    my $t3 = Text::SimpleTable->new( '3', 20, 15 );

    my $cnt = 1;
    for my $id ( @{ $self->{ids} } ) {
        $t3->row( $cnt, $id, $self->{scores}{ $id } );
        $cnt++;
    }
    print $t3->draw;

    if( scalar @{ $self->{filters} } ) {
        my $t4 = Text::SimpleTable->new( 44 );
        $t4->row( 'Filter Users' );
        print $t4->draw;

        my $t5= Text::SimpleTable->new( 3, 20, 15 );
        my $cnt = 1;
        for my $filter ( @{ $self->{filters} } ) {
            $t5->row( $cnt , $filter->{id} , $filter->{score} );
            $cnt++;
        }
        print $t5->draw;
    }

}

sub follow {
use Data::Dumper;
    my $self = shift;
    my $twit = Net::Twitter->new( username=>$self->{setting}{username} ,password=> $self->{setting}{password});
    foreach my $id (  @{ $self->{ids} } ) {
        my $result = $twit->follow( $id );
        sleep( $self->{setting}{sleep} ) if  defined $self->{setting}{sleep};

        my $res = defined $result ? 'ok' : 'fail';
        print "follow $id [$res]\n" if $self->{setting}{on_echo};
    }
}

1;

=head1 NAME

Net::Twitter::FriendFinder - find your twitter friend :-)

=head1 DESCRIPTION

Hello. I started twitter but I did not have much friends.  Since I am shay, so that I created this
module.

You can find twitter friends by using keyword search and then you can follow friends with this module.

=head1 SYNOPSYS

 use Net::Twitter::FriendFinder;

 my $tf 
    = Net::Twitter::FriendFinder->new({ 
        setting => {
            limit => 20,
            on_echo=> 1,
            username => '****',
            password => '****',
            filter_already_followers => 1,
            sleep => 60,
        },
        default => { 
            coverage => 1,
        } ,
        from => {
            Google  => { coverage => 4 },
            Twitter => { handicap => 1.3 } ,
            TwitterKensaku => { coverage => 2 },
            +My::Resource => {}, # your own resource
        }
      });

    # mask who you do not want to find.
    $tf->mask([qw/xxx yyy zzz/]);
    $tf->search( $keyword );
    $tf->show();
    $tf->follow();

    #print Dumepr $twf->ids();
    #print Dumepr $twf->filters();


=head1 MODULE

=head2 new

You should set configulation here.

=over 2 

=item B<setting>

=back

=over 4

=item ->{limit}

you can set limit for how many friends you want to find for max. 

=item ->{on_echo}

when on_echo = 1 then , $tf->follow() method print out who you are going to follow.

=item ->{username}

twitter username

=item ->{password}

twitter password

=item ->{sleep}

since Twitter API has request count limitation, you may want to set sleep time when you try to follow a lot of people.

filter_already_followers - when true, it will check your current followers and not try to follow again. required username and password for this. NOTICE: you may miss some friends to filter who add recently because Twitter API does not return them.

=back

=over 2

=item B<default>

you can set default setting here. 

=back

=over 4


=item ->{coverage}

you can set how much you cover. It depend on where you get resource from , so check the resource POD.

=back

=over 2

=item B<from>

You can set where to search friends from. you can find resource from Net::Twitter::FriendFinder::From::*

  from => {
    Google => {},
    Twitter=> {}
  },


Also if you made your own resource package then set your package name with +.

 from => {
    +My::Resource => { },
 },

=back

=head2 search

this seach method try to search your friends. 

=head2 show

display who you find from your criteria.

=head2 follow

follow people you find. I recommend that before calling this module , you
should check who you are going t follow with show() method.

=head1 HOW TO MAKE YOUR OWN RESOURCE

this is just simple one.

    package My::Recomend::User;
 
    use strict;
    use warnings;

    use base qw/Net::Twitter::FriendFinder::From/;

    sub search {
        return {
            'tomyhero' => 100,
            'perl'     => 30,
        };
    }

    1;

then

    my $tw = Net::Twitter::FriendFinder->new({
        from => {
            +My::Recomend::User => {},
        }
    });

    $tw->search();
    $tw->show();
    $tw->follow();

=head2 mask

you can set screen_names in array ref who you want to avoid search.

 $twf->mask([qw/xxx yyy zzz/]);

=head2 ids

get screen_names

=head2 filters

get filterd screen_names 

=head1 SEE ALSO

L<Net::Twitter>

L<Net::Twitter::From>

L<Net::Twitter::FriendFinder::From::Google>

L<Net::Twitter::FriendFinder::From::Twitter>

L<Net::Twitter::FriendFinder::From::TwitterKensaku>

L<Net::Twitter::FriendFinder::From::TwitterDiff>

L<Net::Twitter::FriendFinder::From::URL>

=head1 AUTHOR

Tomohiro Teranishi<tomohiro.teranishi@gmail.com>

=cut
